File: puzzles/medium/kotlin/src/telephone-numbers.kt
import java.util.Scanner

fun main(args : Array<String>) {
  val input = Scanner(System.`in`)
  val n = input.nextInt()
  var nbNodes = 0
  val root = Node(-1)
  val zero = '0'.toInt()

  for (i in 0 until n) {
    val telephone = input.next()
    var current = root
    for (char in telephone) {
      val digit = char.toInt() - zero
      var index = current.index(digit)
      if (index == -1) {
        current.addChild(digit)
        nbNodes += 1
      }
      index = current.index(digit)
      current = current.children[index]
    }
  }

  // The number of elements (referencing a number) stored in the structure.
  println(nbNodes)
}

class Node(val digit: Int) {
  val children = ArrayList<Node>()

  fun index(d: Int): Int {
    return children.indexOfFirst { it.digit == d }
  }

  fun addChild(d: Int) {
    val child = Node(d)
    children.add(child)
  }
}

File: puzzles/medium/php/telephone-numbers.php
<?php
class Node {
  public $digit;
  public $children = [];

  public function __construct($digit) {
    $this->digit = $digit;
  }

  public function indexOf($digit) {
    for ($i = 0; $i < count($this->children); ++$i) {
      $child = $this->children[$i];
      if ($child->digit == $digit)
        return $i;
    }
    return -1;
  }

  public function addChild($digit) {
    $child = new Node($digit);
    $this->children[] = $child;
    return count($this->children) - 1;
  }
}

$answer = 0; // The number of elements (referencing a number) stored in the structure.
$root = new Node(-1);

fscanf(STDIN, "%d", $count);
for ($i = 0; $i < $count; ++$i) {
  fscanf(STDIN, "%s", $telephone);
  $current = $root;
  for ($j = 0; $j < strlen($telephone); ++$j) {
    $digit = $telephone[$j];
    $index = $current->indexOf($digit);
    if ($index == -1) {
      $index = $current->addChild($digit);
      ++$answer;
    }
    $current = $current->children[$index];
  }
}

echo("$answer\n");
?>
